addons/misra.py

```diff
@@ -2717,34 +2735,53 @@ def misra_14_1(self, data):
 
     def misra_14_2(self, data):
         for token in data.tokenlist:
-            expressions = getForLoopExpressions(token)
-            if not expressions:
-                continue
-            if expressions[0] and not expressions[0].isAssignmentOp:
-                self.reportError(token, 14, 2)
-            elif hasSideEffectsRecursive(expressions[1]):
-                self.reportError(token, 14, 2)
-
-            # Inspect modification of loop counter in loop body
-            counter_vars = getForLoopCounterVariables(token)
-            outer_scope = token.scope
-            body_scope = None
-            tn = token.next
-            while tn and tn.next != outer_scope.bodyEnd:
-                if tn.scope and tn.scope.nestedIn == outer_scope:
-                    body_scope = tn.scope
-                    break
-                tn = tn.next
-            if not body_scope:
-                continue
-            tn = body_scope.bodyStart
-            while tn and tn != body_scope.bodyEnd:
-                if tn.variable and tn.variable in counter_vars:
-                    if tn.next:
-                        # TODO: Check modifications in function calls
-                        if hasSideEffectsRecursive(tn.next):
-                            self.reportError(tn, 14, 2)
-                tn = tn.next
+            if token.str == 'for':
+                expressions = getForLoopExpressions(token)
+                for_clauses = getForLoopClauses(token)
+                if not expressions:
+                    continue
+                if expressions[0] and not expressions[0].isAssignmentOp:
+                    self.reportError(token, 14, 2)
+                if countSideEffectsRecursive(expressions[1]) > 0:
+                    self.reportError(token, 14, 2)
+                if countSideEffectsRecursive(expressions[2]) > 1:
+                    self.reportError(token, 14, 2)
+
+                # Inspect modification of loop counter in loop body
+                counter_vars_first_clause, counter_vars_exit_modified, vars_exit = getForLoopCounterVariables(token)
+                if len(counter_vars_exit_modified) == 0:
+                    # if it's not possible to identify a loop counter, all 3 clauses must be empty
```

Comment:
Isn't getForLoopExpressions enough to determine if an expression is empty. I would expect `expressions[1]` is `None` if there is no condition expression for instance?

Reply:
For the moment, yes
I tried to handle this case https://github.com/danmar/cppcheck/blob/4ff7965c7ed31ed8f367252451a3eb072c4228ab/addons/test/misra/misra-test.c#L1176
and felt like having a function which gives the list of tokens   to handle wierd case like:
`for (j=0, init(&i);i < 10; ++i)` or `for (init(init(&i));i < 10; ++i) // int * init(int *)`
would be more practical. Removing for now as it's not useful